FreeRTOS-Plus/Source/FreeRTOS-IoT-Libraries-LTS-Beta2/c_sdk/platform/freertos/mbedtls/mbedtls_platform_freertos.c

```diff
@@ -0,0 +1,186 @@
+            }
+        }
+    }
+
+    return pBuffer;
+}
+
+/*-----------------------------------------------------------*/
+
+void mbedtls_platform_free( void * ptr )
+{
+    vPortFree( ptr );
+}
+
+/*-----------------------------------------------------------*/
+
+int mbedtls_platform_send( void * ctx,
+                           const unsigned char * buf,
+                           size_t len )
+{
+    Socket_t socket = ctx;
+
+    return ( int ) FreeRTOS_send( socket, buf, len, 0 );
+}
+
+/*-----------------------------------------------------------*/
+
+int mbedtls_platform_recv( void * ctx,
+                           unsigned char * buf,
+                           size_t len )
+{
+    Socket_t socket = ctx;
+
+    return ( int ) FreeRTOS_recv( socket, buf, len, 0 );
+}
+
+/*-----------------------------------------------------------*/
+
+void mbedtls_platform_mutex_init( mbedtls_threading_mutex_t * pMutex )
+{
+    /* Create a statically-allocated FreeRTOS mutex. This should never fail as
+     * storage is provided. */
+    pMutex->mutexHandle = xSemaphoreCreateMutexStatic( &( pMutex->mutexStorage ) );
+    configASSERT( pMutex->mutexHandle != NULL );
+}
+
+/*-----------------------------------------------------------*/
+
+void mbedtls_platform_mutex_free( mbedtls_threading_mutex_t * pMutex )
+{
+    /* Nothing needs to be done to free a statically-allocated FreeRTOS mutex.
+     */
+    ( void ) pMutex;
+}
+
+/*-----------------------------------------------------------*/
+
+int mbedtls_platform_mutex_lock( mbedtls_threading_mutex_t * pMutex )
+{
+    BaseType_t mutexStatus = 0;
+
+    /* mutexStatus is not used if asserts are disabled. */
+    ( void ) mutexStatus;
+
+    /* This function should never fail if the mutex is initialized. */
+    mutexStatus = xSemaphoreTake( pMutex->mutexHandle, portMAX_DELAY );
+    configASSERT( mutexStatus == pdTRUE );
+
+    return 0;
+}
+
+/*-----------------------------------------------------------*/
+
+int mbedtls_platform_mutex_unlock( mbedtls_threading_mutex_t * pMutex )
+{
+    BaseType_t mutexStatus = 0;
+
+    /* mutexStatus is not used if asserts are disabled. */
+    ( void ) mutexStatus;
+
+    /* This function should never fail if the mutex is initialized. */
+    mutexStatus = xSemaphoreGive( pMutex->mutexHandle );
+    configASSERT( mutexStatus == pdTRUE );
+
+    return 0;
+}
+
+/*-----------------------------------------------------------*/
+
+int mbedtls_platform_entropy_poll( void * data,
+                                   unsigned char * output,
+                                   size_t len,
+                                   size_t * olen )
+{
+    int status = 0;
+    NTSTATUS rngStatus = 0;
+
+    /* Context is not used by this function. */
+    ( void ) data;
+
+    /* TLS requires a secure random number generator; use the RNG provided
+     * by Windows. This function MUST be re-implemented for other platforms. */
```

Comment:
Which Windows platform API is being used here for the random number generator?

Reply:
https://docs.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptgenrandom, this is the API being used here.